### PR TITLE
feat(collections): Add `associatewith`

### PR DESCRIPTION
**Pull request by getspooky**
Builds a new Record using the given array as keys and choosing a value for each key using the given selector.

```ts
import { associateWith } from "https://deno.land/std@$STD_VERSION/collections/mod.ts";
import { assertEquals } from "https://deno.land/std@$STD_VERSION/testing/asserts.ts";

const names = ["Kim", "Lara", "Jonathan"];
const namesToLength = associateWith(names, (it) => it.length);

assertEquals(namesToLength, {
  "Kim": 3,
  "Lara": 4,
  "Jonathan": 8,
});
```

Related https://github.com/denoland/deno_std/issues/1173

### CODE REVIEW

File: collections/associate_with.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * const names = [ 'Kim', 'Lara', 'Jonathan' ]
+ * const namesToLength = associateWith(names, it => it.length)
+ *
+ * assertEquals(namesToLength, {
+ *   'Kim': 3,
+ *   'Lara': 4,
+ *   'Jonathan': 8,
+ * })
+ *
```
Comment (LionC) [inline collections/associate_with.ts:21]: That seems to be an accidental additional empty line

File: collections/associate_with.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * Builds a new Record using the given array as keys and choosing a value for each
+ * key using the given selector.
+ *
```
Comment (LionC) [inline collections/associate_with.ts:6]: We should document what happens if the same key is processed multiple times (e.g. with `["a", "b", "a"]` input).

File: collections/associate_with_test.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+import { assertEquals } from "../testing/asserts.ts";
+import { associateWith } from "./associate_with.ts";
+
```
Comment (LionC) [inline collections/associate_with_test.ts:5]: We should test for consistent behaviour if the same key appears mumtiple times in the input (see my comment on docs)

**Review by LionC — CHANGES_REQUESTED** [review]
Thanks for the implementation! I left some comments :-)

File: collections/associate_with.ts (hunk 1 above)
Comment (getspooky) [reply collections/associate_with.ts]: Fixed

File: collections/associate_with_test.ts (hunk 3 above)
Comment (getspooky) [reply collections/associate_with_test.ts:5]: Fixed, @LionC can you please review.

**Review by kt3k — APPROVED** [review]
@getspooky Nice work! LGTM

**Comment by kt3k** [comment]
I'll wait a while before merging to see if there's any inputs, opinions on this because this item looks comparatively new in the proposed list.

**Comment by getspooky** [comment]
I think something went wrong with test :
```sh
error: TS2551 [ERROR]: Property 'Signal' does not exist on type 'typeof Deno'. 'Deno.Signal' is an unstable API. Did you forget to run with the '--unstable' flag, or did you mean 'signal'?

```

**Comment by kt3k** [comment]
@getspooky That CI error is going to be resolved in #1211, but we can't merge that fix right now because that will break stable CI. So let's just ignore canary CI for now.

**Comment by LionC** [comment]
@kt3k can we merge this before the upcoming release?

**Comment by kt3k** [comment]
@LionC Thank you for the ping! Let's merge it.

@getspooky Thanks again for the work!

**Pull request merged**